src/main/scala/com/fulcrumgenomics/bam/OverlappingBasesConsensusCaller.scala

```diff
@@ -0,0 +1,250 @@
+  *
+  * @param onlyMaskDisagreements if the read and mate bases disagree at a given reference position, true to mask (make
+  *                              'N') the read and mate bases, otherwise pick the base with the highest base quality and
+  *                              return a base quality that's the difference between the higher and lower base quality.
+  * @param maxQualOnAgreement    if the read and mate bases agree at a given reference position, true to for the
+  *                              resulting base quality to be the maximum base quality, otherwise the sum of the base
+  *                              qualities.
+  */
+class OverlappingBasesConsensusCaller(onlyMaskDisagreements: Boolean = false,
+                                      maxQualOnAgreement: Boolean = false) {
+  private val NoCall: Byte = 'N'.toByte
+  private val NoCallQual: PhredScore = PhredScore.MinValue
+
+  private val r1BasesBuilder = Array.newBuilder[Byte]
+  private val r1QualsBuilder = Array.newBuilder[PhredScore]
+  private val r2BasesBuilder = Array.newBuilder[Byte]
+  private val r2QualsBuilder = Array.newBuilder[PhredScore]
+  private val r1Builders     = Seq(r1BasesBuilder, r1QualsBuilder)
+  private val r2Builders     = Seq(r2BasesBuilder, r2QualsBuilder)
+
+  /** Consensus calls the overlapping bases if and only if the template is a paired end where both ends map with at
+    * least one base overlapping.
+    *
+    * @param template the template to potentially correct.
+    * @return summary statistics about how many bases were examined and modified
+    */
+  def call(template: Template): CorrectionStats = (template.r1, template.r2) match {
+    case (Some(r1), Some(r2)) if r1.mapped && r2.mapped && r1.matesOverlap.contains(true) => call(r1, r2)
+    case _ => CorrectionStats(0, 0, 0, 0)
+  }
+
+  /** Consensus calls the overlapping bases if and only if both ends map with at least one base overlapping.
+    *
+    * @param r1 the first read in the pair
+    * @param r2 the second read in the pair
+    * @return summary statistics about how many bases were examined and modified
+    */
+  def call(r1: SamRecord, r2: SamRecord): CorrectionStats = {
+    require(r1.mapped && r2.mapped && r1.paired && r2.paired && r1.name == r2.name && r1.matesOverlap.contains(true))
+
+    // Clear and resize the builders
+    r1Builders.foreach { builder =>
+      builder.clear()
+      builder.sizeHint(r1.length)
+    }
+    r2Builders.foreach { builder =>
+      builder.clear()
+      builder.sizeHint(r2.length)
+    }
+
+    val r1Bases = r1.bases
+    val r1Quals = r1.quals
+    val r2Bases = r2.bases
+    val r2Quals = r2.quals
+
+    // Initialize the iters
+    val r1Iter        = MateOverlappingReadAndRefPosIterator(r1, r2).buffered
+    val r2Iter        = MateOverlappingReadAndRefPosIterator(r2, r1).buffered
+    var r1LastReadPos = r1Iter.head.read - 1
+    var r2LastReadPos = r2Iter.head.read - 1
```

Comment:
Seeing this before seeing I was very confused by what `.read` is.  I would rename to `readPos` and `refPos` to make it clearer.